src/MongoDB/Driver/Manager.cpp

```diff
@@ -522,6 +534,136 @@ static bool hippo_mongo_driver_manager_apply_ssl_opts(mongoc_client_t *client, c
+	BsonToVariantConverter convertor(bson_get_data(b_reply), b_reply->len, options);
+	convertor.convert(&v_reply);
+	o_event->o_set(s_MongoDriverMonitoringCommandSucceededEvent_reply, v_reply, s_MongoDriverMonitoringCommandSucceededEvent_className);
+
+	TypedValue args[1] = { *(Variant(o_event)).asCell() };
+	
+	g_context->invokeFuncFew(
+		result.asTypedValue(),
+		m, obj_context,
+		nullptr, 1, args
+	);
+}
+
+
+static void command_failed(const mongoc_apm_command_failed_t *event)
+{
+	ObjectData *obj_context = (ObjectData*) mongoc_apm_command_failed_get_context(event);
+	Class *cls = obj_context->getVMClass();
+	static Class *c_event;
+	Func  *m   = cls->lookupMethod(s_MongoDBDriverManager_dispatchCommandFailed.get());
+	Variant    result;
+	MongoDBDriverManagerData* data = Native::data<MongoDBDriverManagerData>(obj_context);
+
+	Array subscribers = obj_context->o_get(s_MongoDBDriverManager_subscribers, false, s_MongoDriverManager_className).toArray();
+
+	if (subscribers.size() < 1) {
+		return;
+	}
+
+	c_event = Unit::lookupClass(s_MongoDriverMonitoringCommandFailedEvent_className.get());
+	Object o_event = Object{c_event};
+	o_event->o_set(s_MongoDriverMonitoringCommandEvent_commandName, Variant(mongoc_apm_command_failed_get_command_name(event)), s_MongoDriverMonitoringCommandEvent_className);
+	o_event->o_set(s_MongoDriverMonitoringCommandEvent_server, Variant(hippo_mongo_driver_server_create_from_id(data->m_client, mongoc_apm_command_failed_get_server_id(event))), s_MongoDriverMonitoringCommandEvent_className);
+	o_event->o_set(s_MongoDriverMonitoringCommandEvent_operationId, Variant(mongoc_apm_command_failed_get_operation_id(event)), s_MongoDriverMonitoringCommandEvent_className);
+	o_event->o_set(s_MongoDriverMonitoringCommandEvent_requestId, Variant(mongoc_apm_command_failed_get_request_id(event)), s_MongoDriverMonitoringCommandEvent_className);
+	o_event->o_set(s_MongoDriverMonitoringCommandResultEvent_durationMicros, Variant(mongoc_apm_command_failed_get_duration(event)), s_MongoDriverMonitoringCommandResultEvent_className);
+
+	/* Add the error / exception */
+	bson_error_t b_error;
+	mongoc_apm_command_failed_get_error(event, &b_error);
+	o_event->o_set(s_MongoDriverMonitoringCommandFailedEvent_error, MongoDriver::Utils::throwExceptionFromBsonError(&b_error), s_MongoDriverMonitoringCommandFailedEvent_className);
+
+	TypedValue args[1] = { *(Variant(o_event)).asCell() };
+	
+	g_context->invokeFuncFew(
+		result.asTypedValue(),
+		m, obj_context,
+		nullptr, 1, args
+	);
+}
+
+static void hippo_mongo_driver_manager_set_monitoring_callbacks(mongoc_client_t *client, const ObjectData* this_)
+{
+	mongoc_apm_callbacks_t *callbacks = mongoc_apm_callbacks_new();
+
+	mongoc_apm_set_command_started_cb(callbacks, command_started);
+	mongoc_apm_set_command_succeeded_cb(callbacks, command_succeeded);
+	mongoc_apm_set_command_failed_cb(callbacks, command_failed);
+	mongoc_client_set_apm_callbacks(client, callbacks, (void *) this_);
```

Comment:
Possible memory leak here. `_mongoc_client_set_apm_callbacks_private()` (called from `mongoc_client_set_apm_callbacks()`) does a `memcpy()` of its `callbacks` argument to an internal struct. According to the `mongoc_apm_callbacks_new()` documentation, you're expected manually free the allocation with `mongoc_apm_callbacks_destroy()`.